src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,368 @@
+            )
+            module.forward = new_layer.forward  # type: ignore[method-assign]
+            total_swapped += 1
+        elif isinstance(module, lora.Embedding):
+            new_layer = XLoRAEmbeddingLayer(
+                model=xloramodel,
+                target=module,
+                target_forward=module.forward,
+                layer_number=total_swapped,
+                config=config,
+            )
+            module.forward = new_layer.forward  # type: ignore[method-assign]
+            total_swapped += 1
+        elif isinstance(module, lora.Conv2d):
+            new_layer = XLoRAConv2dLayer(
+                model=xloramodel,
+                target=module,
+                target_forward=module.forward,
+                layer_number=total_swapped,
+                config=config,
+            )
+            module.forward = new_layer.forward  # type: ignore[method-assign]
+            total_swapped += 1
+
+    return total_swapped
+
+
+class XLoraModel(LoraModel):
+    """
+    Creates an X-LoRA (Mixture of LoRA experts), model from a pretrained transformers model. Currently,
+    this X-LoRA implementation only works with models with a transformer architecture.
+
+    The method is described in detail in https://arxiv.org/abs/2402.07148.
+
+    Args:
+        model ([`torch.nn.Module`]): The model to be adapted.
+        config ([`XLoraConfig`]): The configuration of the Lora model.
+        adapter_name (`str`): The name of the adapter, does not affect the LoRA adapter names.
+
+    Returns:
+        `torch.nn.Module`: The X-LoRA model.
+
+    Example:
+        ```py
+        >>> from transformers import AutoModelForCausalLM, AutoConfig
+        >>> from peft import LoraConfig, PeftModel, get_peft_model, prepare_model_for_int8_training
+
+        >>> model_config = AutoConfig.from_pretrained("mistralai/Mistral-7B-Instruct-v0.1")
+        >>> config = XLoraConfig(
+        ...     task_type="CAUSAL_LM",
+        ...     hidden_size=model_config.hidden_size,
+        ...     xlora_depth=4,
+        ...     adapters={
+        ...         "adapter_1": "./path/to/the/checkpoint/",
+        ...         "adapter_2": "./path/to/the/checkpoint/",
+        ...         "adapter_n": "./path/to/the/checkpoint/",
+        ...     },
+        ... )
+
+        >>> model = AutoModelForCausalLM.from_pretrained(
+        ...     "mistralai/Mistral-7B-Instruct-v0.1",
+        ...     trust_remote_code=True,
+        ...     use_flash_attention_2=False,
+        ...     device_map="cuda:0",
+        ...     torch_dtype=torch.bfloat16,
+        ... )
+        >>> model = prepare_model_for_int8_training(model)
+        >>> xlora_model = get_peft_model(model, config)
+        ```
+    """
+
+    def __init__(
+        self,
+        model: nn.Module,
+        config: Union[dict[str, XLoraConfig], XLoraConfig],
+        adapter_name: str,
+    ) -> None:
+        super().__init__(model, config, adapter_name)
```

Comment:
From my understanding, XLoRA wouldn't really work with multiple adapters, would it? For instance, below we set `self.xlora_config = peft_config`, which implies that there can only ever be a single XLoRA adapter. If that is so, let's add a check to prevent this. I would suggest to add the check to `inject_adapter`. So add this method here:

```python
    def inject_adapter(self, model: nn.Module, adapter_name: str):
        # XLoRA only supports a single adapter
        if len(set(self.peft_config.keys()) | {adapter_name}) > 1:
            raise ValueError("Trying to add a second adapter, but {self.__class__.__name__} only supports a single adapter")
        return super().inject_adapter(model, adapter_name=adapter_name)
```

Reply:
Thanks for catching that! I have added the method to XLoraModel.